Spigot-Server-Patches/0616-add-get-set-biome-for-namespaced-keys.patch

```diff
@@ -0,0 +1,185 @@
+ 
+         if (iregistry1 == null) {
+@@ -199,14 +200,12 @@ public abstract class IRegistryCustom {
+         }
+ 
+         private static <E> RegistryMaterials<?> d(ResourceKey<? extends IRegistry<?>> resourcekey) {
+-            return new RegistryMaterials<>(resourcekey, Lifecycle.stable());
++            return new RegistryMaterials(resourcekey, Lifecycle.stable()); // Paper - decompile fix
+         }
+ 
+         @Override
+         public <E> Optional<IRegistryWritable<E>> a(ResourceKey<? extends IRegistry<E>> resourcekey) {
+-            return Optional.ofNullable(this.b.get(resourcekey)).map((registrymaterials) -> {
+-                return registrymaterials;
+-            });
++            return Optional.ofNullable((IRegistryWritable<E>) this.b.get(resourcekey)); // Paper - decompile fix
+         }
+     }
+ 
+diff --git a/src/main/java/net/minecraft/server/WorldServer.java b/src/main/java/net/minecraft/server/WorldServer.java
+index 5b0b6edfa790918e56399ff6c83f3feb6e5aca49..26ecdd39fc30baff0f58eb5b212a179e0bb8cd75 100644
+--- a/src/main/java/net/minecraft/server/WorldServer.java
++++ b/src/main/java/net/minecraft/server/WorldServer.java
+@@ -1768,6 +1768,7 @@ public class WorldServer extends World implements GeneratorAccessSeed {
+     public IRegistryCustom r() {
+         return this.server.getCustomRegistry();
+     }
++    public IRegistryCustom getCustomRegistry() { return r(); } // Paper - OBFHELPER
+ 
+     public WorldPersistentData getWorldPersistentData() {
+         return this.getChunkProvider().getWorldPersistentData();
+diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+index b71316cce3bdbf3485be456f0260c6b3463cff8e..bd19397121edabca0d8e5c0bf10bf18cc65476c8 100644
+--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
++++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+@@ -103,6 +103,7 @@ import org.bukkit.Effect;
+ import org.bukkit.FluidCollisionMode;
+ import org.bukkit.GameRule;
+ import org.bukkit.Location;
++import org.bukkit.NamespacedKey;
+ import org.bukkit.Particle;
+ import org.bukkit.Raid;
+ import org.bukkit.Sound;
+@@ -1087,6 +1088,28 @@ public class CraftWorld implements World {
+         return CraftBlock.biomeBaseToBiome(getHandle().r().b(IRegistry.ay), this.world.getBiome(x >> 2, y >> 2, z >> 2));
+     }
+ 
++    // Paper start
++    @Override
++    public NamespacedKey getBiomeKey(int x, int y, int z) {
++        return CraftBlock.biomeBaseToKey(getHandle().getCustomRegistry().getRegistry(IRegistry.getBiomeKeys()), this.world.getBiome(x >> 2, y >> 2, z >> 2));
++    }
++
++
++    @Override
++    public void setBiome(int x, int y, int z, NamespacedKey key) {
++        BiomeBase base = CraftBlock.biomeToBiomeBase(getHandle().getCustomRegistry().getRegistry(IRegistry.getBiomeKeys()), key);
++        BlockPosition pos = new BlockPosition(x, 0, z);
++        if (this.world.isLoaded(pos) && base != null) {
```

Comment:
if base is null, this is an invalid biome: throw an exception

Reply:
ok, if base is null, it will now throw an NPE. I see that CraftWorld uses both Validate and just throws exceptions in different places so I just went with the second method.